sources/packages/backend/apps/workers/src/services/disbursement-schedule/disbursement-schedule-service.ts

```diff
@@ -35,18 +93,130 @@ export class DisbursementScheduleService extends RecordDataModelService<Disburse
+      // Save the disbursements to DB.
+      const disbursementSchedules: DisbursementSchedule[] = [];
+      for (const disbursement of disbursements) {
+        const newDisbursement = new DisbursementSchedule();
+        newDisbursement.creator = auditUser;
+        newDisbursement.disbursementDate = disbursement.disbursementDate;
+        newDisbursement.negotiatedExpiryDate =
+          disbursement.negotiatedExpiryDate;
+        newDisbursement.disbursementValues = disbursement.disbursements.map(
+          (disbursementValue) => {
+            const newValue = new DisbursementValue();
+            newValue.valueType = disbursementValue.valueType;
+            newValue.valueCode = disbursementValue.valueCode;
+            newValue.valueAmount = disbursementValue.valueAmount.toString();
+            newValue.creator = auditUser;
+            return newValue;
+          },
+        );
+        disbursementSchedules.push(newDisbursement);
+      }
+      assessment.disbursementSchedules = disbursementSchedules;
+      assessment.modifier = auditUser;
+      // Save the disbursements.
+      const studentAssessmentRepo =
+        transactionEntityManager.getRepository(StudentAssessment);
+      await studentAssessmentRepo.save(assessment);
+      // Adjust the saved grants disbursements with the values already disbursed.
+      // !Intended to process only grants (CanadaGrant/BCGrant)
+      this.applyGrantsAlreadyDisbursedValues(
+        disbursementSchedules,
+        totalAlreadyDisbursedValues,
+      );
+      // Adjust the saved loans disbursements with the values already disbursed
+      // and generate possible overawards.
+      // !Intended to process only loans (CanadaLoan/BCLoan)
+      await this.applyLoansOverawards(
+        assessment.id,
+        assessment.application.student.id,
+        disbursementSchedules,
+        totalAlreadyDisbursedValues,
+        transactionEntityManager,
+      );
+      // Calculate BC Total grant to be used in e-Cert files.
+      await this.calculateBCTotalGrants(disbursementSchedules);
+      // Persist changes for grants and loans.
+      await studentAssessmentRepo.save(assessment);
+
+      return assessment.disbursementSchedules;
+    });
+  }
+
+  /**
+   * Get the student assessment information needed to process
+   * the disbursements received to be saved.
+   * @param assessmentId student assessment id.
+   * @param entityManager used to execute the commands in the same transaction.
+   * @returns student assessment information.
+   */
+  private async getStudentAssessment(
+    assessmentId: number,
+    entityManager: EntityManager,
+  ): Promise<StudentAssessment> {
+    const studentAssessmentRepo =
+      entityManager.getRepository(StudentAssessment);
+    // Lock the assessment record for update.
+    await studentAssessmentRepo.findOne({
+      select: { id: true },
+      where: { id: assessmentId },
+      lock: { mode: "pessimistic_write" },
```

Comment:
what is pessimistic_write and when is it released

Reply:
Just as information, it is not the first time that we use it. We use the same approach, for instance, for sims.sequence_control table.
It will generate a "FOR UPDATE" for Postgres.
It will be released when the transaction is done.
At the beginning of this transaction, we also set an "idle transaction timeout" to ensure that if something really bad happens it will release the record.
The record is supposed to allow "reads", but if a second query is executed using the "FOR UPDATE" it will wait.
It means if the Zeebe workers are called twice, the first one will execute the task and the second one will wait.